### PR TITLE
[MAINTENANCE] `mypy` typing for `core/util.py`

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
- Typing for `core/util.py`
- `pyproject.toml` has been updated

### CODE REVIEW

**Comment by Kilo59** [comment]
We probably want to use a `TypeVar` and/or `overload` to help narrow the return types of the json conversion function.
When I have a chance I'll propose this change.

**Comment by Shinnnyshinshin** [comment]
> We probably want to use a `TypeVar` and/or `overload` to help narrow the return types of the json conversion function. When I have a chance I'll propose this change.

That would help a ton. Please let me know

File: great_expectations/core/util.py [hunk 1]
```diff
@@ -162,7 +181,10 @@ def determine_progress_bar_method_by_environment() -> Callable:
     return tqdm
 
 
-def convert_to_json_serializable(data) -> dict:  # noqa: C901 - complexity 28
+JSONValues: TypeAlias = Union[dict, list, str, int, float, bool, None]
```
Comment (Kilo59) [inline great_expectations/core/util.py:190]: We might want to consider something like this in a followup.
https://github.com/python/typing/issues/182#issuecomment-1320974824

**Comment by Kilo59** [comment]
This PR should fix the remaining issues

https://github.com/great-expectations/great_expectations/pull/6640

**Review by Kilo59 — APPROVED** [review]
Bit of a self-review here 😄

**Comment by Shinnnyshinshin** [comment]
Thank you @Kilo59

**Comment by Shinnnyshinshin** [comment]
Didn't realize automerge was left on 😓

**Pull request merged**